### ISSUE TITLE
Use dokka with "withJavadocJar()" gradle method

### ISSUE THREAD

**Issue by MrAolen** [Kotlin/dokka#558]
**Question**
Hello,

I would like to know if it is possible to use dokka with the new method "withJavadocJar()" provided by gradle 6 ?

**Installation**
- Operating system: macOS
- Build tool: Gradle v6.0.0
- Dokka version: 0.10.0

**Comment by kamilok1965** [comment]
Not out-of-the-box unfortunately. `withJavadocJar()` packs the output of the `javadoc` task, so you can override this task and run dokka inside

**Comment by elect86** [comment]
@kamildoleglo can you elaborate how would you run dokka inside?

**Comment by AbrahamArmasCordero** [comment]
@elect86 did you found something?

**Comment by elect86** [comment]
nope :/

**Comment by TWiStErRob** [comment]
Ideally it would be as simple as:
```kotlin
project.tasks.named<Jar>("javadocJar") {
	from(tasks.named("dokkaJavadoc"))
}
```
but, sadly `from` is additive, and this above solution only works in Kotlin-only projects, because otherwise Javadoc will be generated and there'll be duplicate files in the JAR. Even setting `duplicateStrategy` doesn't help, because there'll be unnecessary mixed files from JavaDoc and Dokka.

---

The real solution is mimicking what the extension does and creating our own Dokka artifact:
```kotlin
import org.gradle.api.Project
import org.gradle.api.attributes.DocsType
import org.gradle.api.component.AdhocComponentWithVariants
import org.gradle.api.plugins.JavaPlugin
import org.gradle.api.plugins.JavaPluginExtension
import org.gradle.api.plugins.internal.JvmPluginsHelper
import org.gradle.api.tasks.SourceSet
import org.gradle.api.tasks.TaskProvider
import org.gradle.kotlin.dsl.get
import org.gradle.kotlin.dsl.support.serviceOf

/**
 * @see org.gradle.api.plugins.internal.DefaultJavaPluginExtension.withJavadocJar
 */
fun JavaPluginExtension.withDokkaJar(project: Project) {
	JvmPluginsHelper.configureDocumentationVariantWithArtifact(
		JavaPlugin.JAVADOC_ELEMENTS_CONFIGURATION_NAME,
		null,
		DocsType.JAVADOC,
		emptyList(),
		sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME).javadocJarTaskName,
		project.tasks.named("dokkaJavadoc"), // Pick one from org.jetbrains.dokka.gradle.DokkaPlugin.
		project.components["java"] as AdhocComponentWithVariants,
		project.configurations,
		project.tasks,
		project.objects,
		project.serviceOf()
	)
}
```
and then **replace** `project.java.withJavadocJar()` with `project.java.withDokkaJar(project)`.

---

BUT, if you're using `plugins { id("com.gradle.plugin-publish") }`, it automatically calls `withJavadocJar()`, so you cannot replace the call. Luckily it checks if the task already exists  (`tasks.getNames().contains(jarTaskName)`), which means calling `withDokkaJar` BEFORE `project.plugins.apply("com.gradle.plugin-publish")` imperatively kind'a works.

---

@kamildoleglo considering the complexity of this, please reply and expand on your suggestion of "so you can override this task and run dokka inside", we must be missing something trivial 🤞

**Comment by Maragues** [comment]
@TWiStErRob `JvmPluginsHelper.configureDocumentationVariantWithArtifact` no longer exists. There's `configureJavaDocTask`, but I'm not sure it does what we want.

Also, your suggestion doesn't work on android because there's no configuration named `JavaPlugin.JAVADOC_ELEMENTS_CONFIGURATION_NAME`. I tried using `release` or `archives` configuration, but I failed.

Are you aware of a solution for android and modern gradle APIs? Thanks!

**Comment by TWiStErRob** [comment]
It's still there, only its name and signature changed in Gradle 8.1:
https://github.com/TWiStErRob/net.twisterrob.gradle/blob/main/gradle/plugins/src/main/kotlin/net/twisterrob/gradle/build/publishing/withDokkaJar.kt

Good point about Android though, I needed this in a Gradle Plugin, not an AAR.
I diffed `gradlew :foo:dependencies` without and with:
```gradle
android {
	publishing {
		singleVariant("release") {
			withJavadocJar()
		}
	}
}
```
and the thing that popped out was `releaseVariantReleaseJavaDocPublication`
Based on that I searched for `JavaDocPublication"` in Scope=All Places in IDEA and `com.android.build.gradle.internal.dependency.VariantDependenciesBuilder#createLibraryPublishingConfiguration` (AGP 7.4.2) might have the answer.

You probably want to mimic that for Dokka, or thinking about AGP's "Artifact Transformation Pipeline" (dunno the official name), you would probably want to replace `InternalArtifactType.JAVA_DOC_JAR` somehow.
